CIP-plutus-modules/README.md

```diff
@@ -0,0 +1,1643 @@
+there so we can just store the body of the `λ`. When a script is
+evaluated, the value of the script is just added to the environment in
+the same way as the script arguments. The script can then refer to
+its own value using `Self`.
+
+#### Variation: Explicit lambdas
+
+This variation is a less-restrictive version of 'value scripts'. As in
+the former case, we restrict scripts syntactically to explicit
+λ-expressions binding the script arguments, but we do not restrict the
+script body proper to be a syntactic value. As in the former case, the
+λs need not be present in the `Script` representation, because their
+number is known from the number of script arguments, and the bound
+variables are deBruijn indices.
+
+In this variation, script bodies cannot be converted to `CekValue`s
+using `cekValue`; we actually have to run the CEK machine to evaluate
+them. This requires extending the API of the CEK machine, to support
+evaluating a UPLC term *in a given environment*, and returning a
+`CekValue`, rather than a discharged `NTerm`, because discharging a
+`CekValue` loses sharing. Losing sharing is unacceptable because it
+introduces a potentially exponential space cost for acyclic
+structures, and leads to non-termination in the case of cyclic
+structures (created by 'Module-level recursion').
+
+The implementation of the CEK machine currently always discharges
+values before returning them; the core loop of the machine will need
+to be modified to change this.
+
+Since script bodies must be evaluated by running the CEK machine, then
+it is possible to exceed the execution unit budget at any point during
+the script evaluation. The budget must be checked during these
+evaluations, and the budget for evaluating each script will depend on
+the actual costs of evaluating all the previous ones.
+
+To avoid circular dependencies, the scripts must be topologically
+sorted before evaluation, so that no earlier script depends on a later
+one. Topological sorting is linear time in the total number of scripts
+and script arguments.
+
+It is still possible to write a recursive definition of the
+`scriptValues`, so that each script can depend on the *same* map, but
+care is needed to avoid circular dependencies for the reasons
+explained above.
+
+#### A Note on Tuples
+
+The following variations make heavy use of tuples which in practice
+could grow quite large--tuples of modules, and modules as tuples of
+exports. These variations only make sense if projection of a component
+from a tuple is *efficient*, and in particular, constant time,
+independent of the tuple size. At present, tuples are represented
+using the SoP extension (CIP-85) as `constr 0 x1...xn`, but the only
+way to select the `i`th component is using
+```
+  case t of (constr 0 x1...xi...xn) -> xi
+```
+which takes time linear in the size of the tuple to execute, because
+all `n` components need to be extracted from the tuple and passed to
+the case branch (represented by a function).

```

Comment:
Yes that would be logn `case` terms, cheaper in terms of execution units (at least for long tuples) but bigger in script size.

Reply:
Logarithmic is better than linear, but it's also the cost of accessing variables in the environment (which is logarithmic in the size of the environment). So the advantage of putting the module exports into one tuple instead of bunging them all into the environment would disappear. Much better to bite the bullet and put in explicit projections, getting constant time access.